### ISSUE TITLE
Implement control-dependent running costs

### ISSUE THREAD

**Issue by goerz** [JuliaQuantumControl/GRAPE.jl#30]
Assuming `J({ϵₙₗ}) = J_T({|Ψₖ(T)⟩) + J_a({ϵₗ(t)}) + ∫g_b({|Ψ(t)⟩)dt`, we want to add support for the term `J_a` in GRAPE. We do not assume that `J_a = ∫g_a({ϵₗ(t)})dt`.

See https://github.com/JuliaQuantumControl/QuantumControl.jl/issues/15

The user should be able to pass a function `J_a` and potentially `grad_J_a` to the `optimize` routine (respectively to the initializer for a `ControlProblem`). If `grad_J_a` is not given, a `make_grad_J_a` routine will use Zygote to produce `grad_J_a` via automatic differentiation.

The interface for the routines is as follows:

* `J_a(pulsevals, controls, index_map, tlist)` where `pulsevals` is a vector containing the combined pulse amplitudes from all controls, `controls` is a vector of controls (functions, in most cases), `index_map` is an `IdDict` that maps each control to an iterable of indices in `pulsevals`, allowing to select the specific values in `pulsevals` that are associated with a particular control. Lastly, `tlist` is the time grid.

* `grad_J_a!(∇J_a, pulsevals, controls, index_map, tlist)` should put the vector `∂J_a/∂ϵₙ`for all values `ϵₙ` in `pulsevals` into `∇J_a`.

* `make_grad_J_a(J_a, controls, index_map, tlist)` should return a `grad_J_a!` function by internally mapping to `Zygote.gradient(pulsevals -> J_a(...))`

The requirement to be compatible with Zygote motivates the use of the `index_map`, instead of the `parameters` dict that is used for propagation, which maps each control to a view of values in `pulsevals`; I don't think Zygote will deal with aliases: everything must be traceable directly to `pulsevals`.

**Comment by goerz** [comment]
@seba-car in https://github.com/JuliaQuantumControl/QuantumControl.jl/issues/15#issuecomment-1259866332: 
> The proposed interface looks good. It is minimal and consistent with everything.

The question is whether we want to include `controls` and `index_map` as parameters. Strictly for GRAPE, we don't technically need to. There's the implication that the pulse values are the controls discretized to the midpoints of the time grid. So, given just `pulsevals` and `tlist`, a simple `reshape` gives access to the pulse values for a specific control. We currently use that for `J_T`, e.g. in [`grad_J_T_via_chi!`](https://github.com/JuliaQuantumControl/QuantumControlBase.jl/blob/9ac6881eff246deefc12b198f94f283a155ca65a/src/functionals.jl#L466). Right now, we use an interspersed convention `∈ₗₙ` (control index first, time index second), but I might want to switch that around.

We can't get away with a simple `reshape` if we want to also use `J_T`/`J_a` and their gradients more generally, when the control parameters aren't just the pulse amplitudes. For example, when doing GROUP (that's gradient-enhanced CRAB). As long as all controls have the same number of parameters, you could still reshape, but you'd at least need to know the number of controls. It seems unlikely, but at least hypothetically one could have a different number of parameters for different controls. In that case, you can't just reshape, but you'd really need a list of `controls` and an `index_map` as proposed above. Also, interspersing the parameters when combining them into a vector wouldn't make sense at all (which is one reason why I might want to switch from `∈ₗₙ` to `∈ₙₗ`).

The other benefit of using `controls` and `index_map` is that any such convention is no longer implicit.

I'm still a little undecided if I should start out with the simpler version (without `controls` and `index_map`) or do both `J_a` and `J_T` in the most general way (which would require changing the existing interface for final time functionals). If we go with the simpler version, we might have to rewrite both when we get around to GROUP and other generalized gradient-based methods.

**Comment by seba-car** [comment]
I understand that GROUP needs only an extra chain rule with respect to GRAPE. The derivative of the pulse amplitudes with respect to the parameters. Maybe it would make sense to do the simple case using just a `reshape` and then add some method to support parametrization.

**Comment by goerz** [comment]
Fair enough. Also, I realize we could just have both (yay for multiples dispatch!) So let's stick to 

```julia
J_a(pulsevals, tlist)
grad_J_a!(∇J_a, pulsevals, tlist)
make_grad_J_a(J_a, tlist)
```

for now, with an implicit convention for `pulsevals`, and then at some point in the future we can add

```julia
J_a(combined_parameters, controls, index_map, tlist)
grad_J_a!(∇J_a, combined_parameters, controls, index_map, tlist)
make_grad_J_a(J_a, controls, index_map, tlist)
```

whenever we need it.

**Comment by goerz** [comment]
Closed by #32 and https://github.com/JuliaQuantumControl/QuantumControlBase.jl/pull/43
